File: leetcode/278.first-bad-version.py
# @leet start
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:


class Solution:
    def firstBadVersion(self, n: int) -> int:
        # Do a variant of a binary search
        i = 1
        j = n

        while i < j:
            mid = (i + j) // 2

            if isBadVersion(mid):
                j = mid
            else:
                i = mid + 1

        return i


# @leet end

File: leetcode/278.first-bad-version.ts
// @leet start
/**
 * The knows API is defined in the parent class Relation.
 * isBadVersion(version: number): boolean {
 *     ...
 * };
 */

const solution = function (isBadVersion: (version: number) => boolean) {
  return function (n: number): number {
    let lo = 0;
    let hi = n;

    while (lo <= hi) {
      const mid = Math.floor((lo + hi) / 2);

      if (isBadVersion(mid)) {
        hi = mid - 1;
      } else {
        lo = mid + 1;
      }
    }

    return lo;
  };
};
// @leet end
